### PR TITLE
Rename XamlIL to XamlX and refactor out the hard IL dependencies.

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Rename XamlIL -> XamlX. Refactor out base classes for the compilers for 3 built-in compiler types, declarative, imperative, and IL (each builds on previous). Make the backend generic on the "raw" emitter and the result type returned by the node emitters.

Keep support for implementing emit for specific backends on the node types themselves while also supporting using emitter objects exclusively.

Contributes to #8.

### CODE REVIEW

**Comment by rcollina** [comment]
I can't seem to find a nuget package for this project. I'd love to use Xaml in .NET Core for x-plat applications.

**Comment by kekekeks** [comment]
@rcollina For now it's supposed to be included as a git submodule since the API isn't really stable at all. You also will have to use ILRepack to get it merged into your MSBuild task anyway.

**Comment by jkoritzinsky** [comment]
@kekekeks can you review this when you have a chance? I'd like to eventually get this merged in so we can update Avalonia and work on sharing more phases.

**Comment by kekekeks** [comment]
@jkoritzinsky I think we need to do a `filter-branch`-based rename and re-apply your refactoring...

**Comment by kekekeks** [comment]
Mkay, in [this](https://github.com/kekekeks/XamlIl/commits/xamlx-rename) universe Oceania had *always* been at war with Eastasia

**Comment by jkoritzinsky** [comment]
I’m good with us doing a filter-branch style renaming.

**Comment by kekekeks** [comment]
Force-pushed to https://github.com/kekekeks/XamlIl/tree/xamlx-rename with commit messages rewritten as well for the greater glory of Ingsoc.

**Comment by jkoritzinsky** [comment]
I highly appreciate the 1984 references haha. Let’s go with your branch.

**Comment by kekekeks** [comment]
Rename utility that was used for `--tree-filter`

```cs
        static int Main(string[] args)
        {
            if (args.Length != 3)
            {
                Console.Error.WriteLine("Usage: app <target> <oldname> <newname>");
                return 1;
            }

            var target = args[0];
            var oldNameString = args[1];
            var newNameString = args[2];

            void Do(string path)
            {
                foreach (var f in Directory.GetFiles(path))
                {
                    var name = Path.GetFileName(f);
                    var newName = name.Replace(oldNameString, newNameString);
                    var newPath = Path.Combine(path, newName);
                    if (name != newName)
                    {
                        Console.WriteLine($"Renaming {f} to {newName}");
                        File.Move(f, newPath);
                    }
                    Console.WriteLine($"Replacing text in {newPath}");
                    File.WriteAllText(newPath, File.ReadAllText(newPath).Replace(oldNameString, newNameString));
                }

                foreach (var d in Directory.GetDirectories(path))
                {
                    var name = Path.GetFileName(d);
                    if(name.StartsWith(".git"))
                        continue;
                    var newName = name.Replace(oldNameString, newNameString);
                    var newPath = Path.Combine(path, newName);
                    if (name != newName)
                    {
                        Console.WriteLine($"Renaming {d} to {newName}");
                        Directory.Move(d, newPath);
                    }

                    Do(newPath);
                }
            }
            Do(target);

            return 0;
        }
```

Might be worth to apply it to PR commits so they could be cherry-picked more easily.

**Comment by jkoritzinsky** [comment]
I'll give it a shot. The command would be `git filter-branch --tree-filter "ConsoleApp.exe XamlIL XamlX"` where ConsoleApp is your utility script correct?

**Comment by kekekeks** [comment]
I was using a separate bash script that looks like 
```
#!/bin/bash
dotnet /path/to/app.dll `pwd` XamlIl XamlX
dotnet /path/to/app.dll `pwd` XamlIL XamlX
```

You can change the script by hardcoding those values and using the current directory instead of the first command line arg.

**Comment by jkoritzinsky** [comment]
Superseded by #11

File: src/XamlX/Emit/XamlEmitContextWithLocals.cs [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        public XamlLocalsPool.PooledLocal GetLocalOfType(IXamlType type)
+        {
+            return Emitter.LocalsPool.GetLocal(type);
+        }
+
+        protected override TEmitResult EmitNodeCore(IXamlAstNode value, TBackendEmitter codeGen, out bool foundEmitter)
+        {
+            var result = base.EmitNodeCore(value, codeGen, out foundEmitter);
+
+            if (result != default)
```
Comment (kekekeks) [inline src/XamlX/Emit/XamlEmitContextWithLocals.cs:56]: Wont even compile with C# 9

File: src/XamlX/Emit/XamlEmitContext.cs [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+
+        protected virtual TEmitResult EmitNodeCore(IXamlAstNode value, TBackendEmitter codeGen, out bool foundEmitter)
+        {
+            TEmitResult res = default;
+            foreach (var e in Emitters)
+            {
+                if (e is IXamlAstNodeEmitter<TBackendEmitter, TEmitResult> ve)
+                {
+                    res = ve.Emit(value, this, codeGen);
+                    if (res != default)
```
Comment (kekekeks) [inline src/XamlX/Emit/XamlEmitContext.cs:199]: Won't even compile with C# 9

**Pull request closed without merging**